### DISCUSSION TITLE
PE checksums on Linux Builds are always 0, intentional?

### DISCUSSION THREAD

**Discussion by joshudson** [dotnet/roslyn#74654] [General]
I finally got around to changing the error message in the DOS Stub (for what it's worth I always hated that particular error message and now that badly written 16 bit assembly bugs me too) in my project; and I discovered something oddball.

After replacing the error message the PE checksum would be wrong so I have to fix that too, so of course it's compute the delta of the checksum and patch it in, right? Nope!

When doing a build via `dotnet build` on Linux, the checksum field is always filled with a 0.

Hmmm, very hmmm.

**Comment by 333fred** [comment]
Are you actually emitting portable or embedded PDBs? In general, this question is extremely vague and lacking in details. If you want help with a specific thing, it's best to provide an actual example, and show what you're doing and what you're expecting to happen, rather than asking us to guess at the details here.

**Comment by joshudson** [comment]
@333fred : The PE checksum is found 64 bytes into the optional header, or 216 bytes into every dll emitted by dotnet build.

I can't imagine how that means anything else.

```
$ hexdump -C hello.dll
00000000  4d 5a 90 00 03 00 00 00  04 00 00 00 ff ff 00 00  |MZ..........ÿÿ..|
00000010  b8 00 00 00 00 00 00 00  40 00 00 00 00 00 00 00  |¸.......@.......|
00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 80 00 00 00  |................|
00000040  0e 1f ba 0e 00 b4 09 cd  21 b8 01 4c cd 21 54 68  |..º..´.Í!¸.LÍ!Th|
00000050  69 73 20 70 72 6f 67 72  61 6d 20 63 61 6e 6e 6f  |is program canno|
00000060  74 20 62 65 20 72 75 6e  20 69 6e 20 44 4f 53 20  |t be run in DOS |
00000070  6d 6f 64 65 2e 0d 0d 0a  24 00 00 00 00 00 00 00  |mode....$.......|
00000080  50 45 00 00 4c 01 03 00  a3 06 94 a9 00 00 00 00  |PE..L...£..©....|
00000090  00 00 00 00 e0 00 22 00  0b 01 30 00 00 08 00 00  |....à."...0.....|
000000a0  00 08 00 00 00 00 00 00  9a 26 00 00 00 20 00 00  |.........&... ..|
000000b0  00 40 00 00 00 00 40 00  00 20 00 00 00 02 00 00  |.@....@.. ......|
000000c0  04 00 00 00 00 00 00 00  04 00 00 00 00 00 00 00  |................|
000000d0  00 80 00 00 00 02 00 00  00 00 00 00 03 00 60 85  |..............`.|
                              Here ^^ ^^ ^^ ^^

$ objdump -x hello.dll | grep CheckSum
CheckSum		00000000
```

**Reply by 333fred** [reply]
Please provide the info I actually asked for 🙂. Again, I can't help by just guessing at things.

**Reply by joshudson** [reply]
It does not matter what PDB format is being used.

You can try this yourself:
```
/tmp$ (mkdir hello && cd hello && dotnet new console && dotnet build && objdump -x bin/*/*/hello.dll | grep CheckSum)
CheckSum		00000000
```

The `objdump` command outputs fields from the PE header, one of which is CheckSum. A checksum of all zeros is definitely unexpected.

**Reply by 333fred** [reply]
Thank you for providing the repro I asked for. I presume, then, that you're not strong-name signing the assembly? The checksum is only used in that scenario. When not strong-name signing, we simply write 0 to the header; the checksum is only relevant when strong-name signing with a private key (ie, not `PublicSign`), as otherwise it can simply be overwritten with no consequences.

**Reply by joshudson** [reply]
Virus scanners have been intermittently picky about PE checksums. This generally results in me complaining to the virus scanner vendor but when somebody else gets to pick the virus scanner and wants to know why it's complaining it gets annoying.
